src/Featurizers/Components/GrainEstimatorImpl.h

```diff
@@ -0,0 +1,642 @@
+///                 a `GrainEstimator`.
+///
+template <typename GrainT, typename EstimatorT>
+struct GrainEstimatorTraits {
+    // ----------------------------------------------------------------------
+    // |  Public Types
+    using InputType                         = std::tuple<GrainT, typename EstimatorT::InputType>;
+    using TransformedType                   = std::tuple<GrainT, typename Details::EstimatorOutputType<EstimatorT>::type>;
+};
+
+/////////////////////////////////////////////////////////////////////////
+///  \class         GrainTransformer
+///  \brief         A Transformer that applies a Transformer unique to the
+///                 observed grain using grain-specific state.
+///
+template <typename GrainT, typename EstimatorT>
+class GrainTransformer :
+    public Transformer<
+        typename GrainEstimatorTraits<GrainT, EstimatorT>::InputType,
+        typename GrainEstimatorTraits<GrainT, EstimatorT>::TransformedType
+    > {
+public:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Types
+    // |
+    // ----------------------------------------------------------------------
+    using TheseGrainEstimatorTraits         = GrainEstimatorTraits<GrainT, EstimatorT>;
+
+    using BaseType =
+        Transformer<
+            typename TheseGrainEstimatorTraits::InputType,
+            typename TheseGrainEstimatorTraits::TransformedType
+        >;
+
+    using GrainTransformerType =
+        Transformer<
+            typename EstimatorT::InputType,
+            typename EstimatorT::TransformedType
+        >;
+
+    using GrainTransformerTypeUniquePtr     = std::unique_ptr<GrainTransformerType>;
+    using TransformerMap                    = std::map<GrainT, GrainTransformerTypeUniquePtr>;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    GrainTransformer(TransformerMap transformers);
+    GrainTransformer(Archive &ar);
+
+    ~GrainTransformer(void) override = default;
+
+    FEATURIZER_MOVE_CONSTRUCTOR_ONLY(GrainTransformer);
+
+    void save(Archive &ar) const override;
+
+private:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Data
+    // |
+    // ----------------------------------------------------------------------
+    TransformerMap                          _transformers;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Methods
+    // |
+    // ----------------------------------------------------------------------
+
+    // MSVC has problems when the declaration and definition are separated
+    void execute_impl(typename BaseType::InputType const &input, typename BaseType::CallbackFunction const &callback) override {

```

Comment:
I figured out more info about why MSVC has issues with this. It has to do with how many times it has to "unwrap" a type to find the actual type. Im sure thats not really the root cause, but it seems like if it has to take more then about 3 steps to find the actual type it doesn't succeed. I was able to repro it on a much smaller scale. We can sync on it some time.